python/dolfinx/fem/function.py

```diff
@@ -473,79 +478,112 @@ def split(self) -> tuple[Function, ...]:
+    :param symmetry: Symmetry option for blocked tensor elements.
+
+    """
     family: str
     degree: int
+    shape: typing.Optional[typing.Tuple[int, ...]] = None
+    symmetry: typing.Optional[bool] = None
+
+
+def FunctionSpace(mesh: Mesh,
+                  element: typing.Union[ufl.FiniteElementBase, ElementMetaData,
+                                        typing.Tuple[str, int, typing.Tuple, bool]],
+                  form_compiler_options: typing.Optional[dict[str, typing.Any]] = None,
+                  jit_options: typing.Optional[dict[str, typing.Any]] = None) -> FunctionSpaceBase:
+    """Create a finite element function space.
+
+    Args:
+        mesh: Mesh that space is defined on
+        element: Finite element description
+        form_compiler_options: Options passed to the form compiler
+        jit_options: Options controlling just-in-time compilation
+
+    Returns:
+        A function space.
+
+    """
+
+    # Create UFL element
+    try:
+        e = ElementMetaData(*element)
+        ufl_e = basix.ufl.element(e.family, mesh.basix_cell(), e.degree, shape=e.shape,
+                                  gdim=mesh.ufl_cell().geometric_dimension())
+    except TypeError:
+        ufl_e = element  # type: ignore
+
+    # Check that element and mesh cell types match
+    if ufl_e.cell() != mesh.ufl_domain().ufl_cell():
+        raise ValueError("Non-matching UFL cell and mesh cell shapes.")
+
+    # Compile dofmap and element and create DOLFIN objects
+    dtype = mesh.geometry.x.dtype
+    if form_compiler_options is None:
+        form_compiler_options = dict()
+    if dtype == np.float32:
+        form_compiler_options["scalar_type"] = "float"
+    elif dtype == np.float64:
+        form_compiler_options["scalar_type"] = "double"
+    else:
+        raise RuntimeError("Unsupported geometry type. Must be float32 or float64.")
+
+    (ufcx_element, ufcx_dofmap), module, code = jit.ffcx_jit(mesh.comm, ufl_e,
+                                                             form_compiler_options=form_compiler_options,
+                                                             jit_options=jit_options)
+    ffi = module.ffi
+    if dtype == np.float32:
+        cpp_element = _cpp.fem.FiniteElement_float32(ffi.cast("uintptr_t", ffi.addressof(ufcx_element)))
+    elif dtype == np.float64:
+        cpp_element = _cpp.fem.FiniteElement_float64(ffi.cast("uintptr_t", ffi.addressof(ufcx_element)))
+    cpp_dofmap = _cpp.fem.create_dofmap(mesh.comm, ffi.cast(
+        "uintptr_t", ffi.addressof(ufcx_dofmap)), mesh.topology, cpp_element)
+
+    # Initialize the cpp.FunctionSpace
+    try:
+        cppV = _cpp.fem.FunctionSpace_float64(mesh._cpp_object, cpp_element, cpp_dofmap)
+    except TypeError:
+        cppV = _cpp.fem.FunctionSpace_float32(mesh._cpp_object, cpp_element, cpp_dofmap)
+
+    return FunctionSpaceBase(mesh, ufl_e, cppV)
 
 
-class FunctionSpace(ufl.FunctionSpace):
+class FunctionSpaceBase(ufl.FunctionSpace):
```

Comment:
I would rather like it if we kept the name `FunctionSpace` for the class, and went directly for `functionspace` as a constructor above. I don't think having a base-class that is never extended makes sense.

Reply:
Eventually I would like to have the function `functionspace` that returns a space, and rename the class `FunctionSpaceBase` -> `FunctionSpace`. But, this is an interface change which I think we might need to wait a release for.

I've now added `functionspace` and add a deprecation not to the function `FunctionSpace`.